numpy/core/shape_base.py

```diff
@@ -275,3 +278,62 @@ def hstack(tup):
         return _nx.concatenate(arrs, 0)
     else:
         return _nx.concatenate(arrs, 1)
+
+def stack(arrays, axis=0):
+    """
+    Join a sequence of arrays along a new axis.
+
+    .. versionadded:: 1.10.0
+
+    Parameters
+    ----------
+    arrays : sequence of ndarrays
+        Each array must have the same shape.
+    axis : int, optional
+        The axis along which the arrays will be stacked.
+
+    Returns
+    -------
+    stacked : ndarray
+        The stacked array has one more dimension than the input arrays.
+
+    See Also
+    --------
+    concatenate : Join a sequence of arrays along an existing axis.
+    split : Split array into a list of multiple sub-arrays of equal size.
+
+    Examples
+    --------
+    >>> arrays = [np.random.randn(3, 4) for _ in range(10)]
+    >>> np.stack(arrays, axis=0).shape
+    (10, 3, 4)
+
+    >>> np.stack(arrays, axis=1).shape
+    (3, 10, 4)
+
+    >>> np.stack(arrays, axis=2).shape
+    (3, 4, 10)
+
+    >>> a = np.array([1, 2, 3])
+    >>> b = np.array([2, 3, 4])
+    >>> np.stack((a, b))
+    array([[1, 2, 3],
+           [2, 3, 4]])
+
+    >>> np.stack((a, b), axis=-1)
+    array([[1, 2],
+           [2, 3],
+           [3, 4]])
+
+    """
+    arrays = [_nx.asarray(arr) for arr in arrays]
```

Comment:
The problem with handling subclasses correctly in general is that it is
literally impossible. Subclasses might have completely arbitrary behavior,
so basically any behavior on the numpy side will be correct in some cases.
Should we preserve subclasses when doing X? Yes, that is the only correct
behavior (sometimes). Should we drop subclasses when doing X? Yes, that is
also required (other times). Should we return a different subclass?
Entirely possible. Use one specific set of indexing operations or function
calls instead of another purportedly equivalent approach? Definitely
necessary. Which specific set? Well, there are contradictory answers. Stand
on our heads? Can't rule it out...

Historically afaict the rule has been that we do whatever hack happens to
work for whichever subclass authors file a bug last...

What I'm saying is, I don't think we should waste too much energy worrying
about such questions because the best we can hope for is to choose who we
annoy :-/
On Feb 25, 2015 9:34 PM, "Stephan Hoyer" <notifications@github.com> wrote:

> In numpy/core/shape_base.py
> <https://github.com/numpy/numpy/pull/5605#discussion_r25405780>:
>
> > +    >>> np.stack(arrays, axis=2).shape
> > +    (3, 4, 10)
> > +
> > +    >>> a = np.array([1, 2, 3])
> > +    >>> b = np.array([2, 3, 4])
> > +    >>> np.stack((a, b))
> > +    array([[1, 2, 3],
> > +           [2, 3, 4]])
> > +
> > +    >>> np.stack((a, b), axis=-1)
> > +    array([[1, 2],
> > +           [2, 3],
> > +           [3, 4]])
> > +
> > +    """
> > +    arrays = [_nx.asarray(arr) for arr in arrays]
>
> @mhvk <https://github.com/mhvk> I'm all for supporting ndarray subclasses
> when feasible, but the options we've come up with so far would either
> entail worse error reporting for non-compliant subclasses (using
> asanyarray) or additions to the function signature (adding subok=False)
> with no actual benefits, given the limitations of concatenate.
>
> Are there any non-trivial ndarray subclasses that *would* benefit from
> these changes? I suppose we could argue that it is best to keep the door
> open to improvements in the future.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5605/files#r25405780>.
>